jina/types/arrays/abstract.py

```diff
@@ -0,0 +1,228 @@
+import re
+from abc import ABC, abstractmethod
+from math import inf
+from typing import Union, List, Tuple, Iterable, Callable, Optional, Dict, Iterator
+
+
+class AbstractDocumentArray(ABC):
+    """ Abstract class that defines the public interface of DocumentArray classes """
+
+    @abstractmethod
+    def get_attributes(self, *fields: str) -> Union[List, List[List]]:
+        """Return all nonempty values of the fields from all docs this array contains
+
+        :param fields: Variable length argument with the name of the fields to extract
+        """
+        pass
+
+    @abstractmethod
+    def get_attributes_with_docs(
+        self,
+        *fields: str,
+    ) -> Tuple[Union[List, List[List]], 'DocumentArray']:
+        """Return all nonempty values of the fields together with their nonempty docs
+
+        :param fields: Variable length argument with the name of the fields to extract
+        """
+        pass
+
+    @abstractmethod
+    def traverse(
+        self, traversal_paths: Iterable[str]
+    ) -> Iterable['TraversableSequence']:
+        """
+        Return an Iterator of :class:``TraversableSequence`` of the leaves when applying the traversal_paths.
+        Each :class:``TraversableSequence`` is either the root Documents, a ChunkArray or a MatchArray.
+
+        :param traversal_paths: a list of string that represents the traversal path
+        """
+        pass
+
+    @abstractmethod
+    def traverse_flat_per_path(
+        self, traversal_paths: Iterable[str]
+    ) -> Iterable['TraversableSequence']:
+        """
+        Returns a flattened :class:``TraversableSequence`` per path in :param:``traversal_paths``
+        with all Documents, that are reached by the path.
+
+        :param traversal_paths: a list of string that represents the traversal path
+        """
+        pass
+
+    @abstractmethod
+    def traverse_flat(self, traversal_paths: Iterable[str]) -> 'TraversableSequence':
+        """
+        Returns a single flattened :class:``TraversableSequence`` with all Documents, that are reached
+        via the :param:``traversal_paths``.
+
+        .. warning::
+            When defining the :param:``traversal_paths`` with multiple paths, the returned
+            :class:``Documents`` are determined at once and not on the fly. This is a different
+            behavior then in :method:``traverse`` and :method:``traverse_flattened_per_path``!
+
+        :param traversal_paths: a list of string that represents the traversal path
+        """
+        pass
+
+    @abstractmethod
+    def match(
```

Comment:
I am not sure `match` is offered by `DocumentArrayMemmap`, if so, make sure is tested

Reply:
yes it's offered and `tests/unit/types/arrays/test_neural_ops.py` now tests various combinations of DA and DAM

Reply:
`NeuralOpsMixin ` is not implemented by `DocumentArrayMemmap`

Reply:
![image](https://user-images.githubusercontent.com/15269265/128691627-4ea9fc4b-28d8-4344-aae7-6a31ddd3c842.png)

I'm confused, why do you think so ?

Reply:
Not in master currently

Reply:
yeah but the purpose of this PR is to make it possible to add neural ops. So `DocumentArrayNeuralOpsMixin` are added to `DocumentArrayMemmap` in this PR

Reply:
@alaeddine-13 is right.